Updated text + added 'copy link' button

svub opened the pull request:
Reason: user got stuck in the "Uh oh, something went wrong" w/t knowing how to continue
Change: 
* Text with more concrete help (using browser names instead of the word "browser")
* A button to easily copy the link
* if trigger `navigator.share()` if available
Goal: help the user to overcome issue

svub committed:
Updating text + adding 'copy link' button

danimoh commented on dist/browser-warning.html.template:
```diff
@@ -1,6 +1,9 @@
-        <h1>Uh oh, something went wrong.</h1>
+        <h1 id="browser-warning-headline">Uh oh, something went wrong.</h1>
+        <div id="browser-warning-copy-link">
+            <pre></pre><button class="nq-button">Copy link</button>
```
What's the `<pre>` and the container for?

svub replied:
Oh, right, the idea was to actually show the URL. Wdyt is better: show the URL or remove those containers and just leave the button?

danimoh commented on dist/browser-warning.js:
```diff
@@ -145,24 +145,75 @@
+            $warningHeadline.textContent = 'Please copy the link and paste it directly in Safari, Chrome or Firefox.';
+            $warningMessage.textContent = 'You\'re currently in a so-called in-app browser.';
+            $warningCopyLink.classList.add('show');
+            $warningCopyLink.getElementsByTagName('button')[0].addEventListener('click', e => {
```
Use a normal `function` instead of an arrow function to avoid that old browsers can't parse this file.

danimoh commented on dist/browser-warning.js:
```diff
@@ -145,24 +145,75 @@
+                copy(location.href);
+            });
+            if (navigator.share) {
+                navigator.share(location.href);
```
See here for the correct api for `navigator.share`: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/share

svub replied:
Thank! Misread the parameter definition.

danimoh commented on dist/browser-warning.js:
```diff
@@ -145,24 +145,75 @@
+        var activeInput = document.activeElement
+            && (document.activeElement.nodeName === 'INPUT' || document.activeElement.nodeName === 'TEXTAREA')
+            ? document.activeElement
+            : null;
```
I think in the context of the browser warning we don't really need to keep the `originalRange` or refocus the `activeInput`. I would remove this logic to not bloat the file unnecessarily.

danimoh commented on dist/browser-warning.js:
```diff
@@ -145,24 +145,75 @@
         var $warningMessage = document.getElementById('browser-warning-message') || $warningContainer;
         if (warning === 'web-view') {
-            $warningMessage.textContent = 'You are viewing this page from inside another app. Please use a real browser.'
+            $warningHeadline.textContent = 'Please copy the link and paste it directly in Safari, Chrome or Firefox.';
```
I would put Safari last.

danimoh commented on dist/browser-warning.js:
```diff
@@ -145,24 +145,75 @@
-            $warningMessage.textContent = 'You are viewing this page from inside another app. Please use a real browser.'
+            $warningHeadline.textContent = 'Please copy the link and paste it directly in Safari, Chrome or Firefox.';
+            $warningMessage.textContent = 'You\'re currently in a so-called in-app browser.';
+            $warningCopyLink.classList.add('show');
```
Instead of having the button already in the markup and requiring css to hide/unhide it, you could just create the button here when required.

danimoh reviewed (changes requested):
![Screenshot](https://user-images.githubusercontent.com/6204514/70760287-c9c37900-1d0e-11ea-8773-833c91058e84.png)
I think the headline is too long. Also the button should probably be the last element.

svub committed:
addressing change requests

sisou commented:
Copying the link into another browser does not work. The Hub requires a referrer (for redirects) or a window reference from the opening website (for popups), both of which get lost when only copy-pasting the link into another browser/tab/instance.

The only solution here is to re-open the opening app (from which the Hub was launched) in another browser, not the Hub itself.

danimoh commented:
The browser warning is included in all our apps that use the hub.
Thus, the user will already get the warning before getting into the hub.
But third party developers might actually have that problem if they don't use our browser warnings.
Best would probably to display a different error if we're in the hub.

danimoh commented on dist/browser-warning.js:
```diff
@@ -146,17 +146,21 @@
+
             if (navigator.share) {
-                navigator.share(location.href);
+                navigator.share({ data: location.href });
```
That's still wrong usage of `navigator.share`.
Also, `navigator.share` only works as a response to a user interaction.
This should thus happen on the button click.

danimoh replied:
I fixed this

danimoh commented on dist/browser-warning.js:
```diff
@@ -182,38 +186,23 @@
         return isSuccess;
     }
 
+    return showWarning('web-view');
```
Doesn't belong in the PR.

danimoh replied:
I fixed this.

danimoh committed:
Use navigator.share instead of copy if supported

If supported by the browser, trigger navigator.share on click of the
share button instead of copying the link and change the messages/hints
accordingly. This fixes the previous wrong use of navigator.share which
used an invalid argument format and was invoked without used interaction
which gets blocked by the browser.

danimoh committed:
Remove demo code

danimoh committed:
Add share button to all warnings and adapt instructions

danimoh committed:
Add onBrowserWarning callback support

Implement a callback that be used to configue the warning to display.

danimoh reviewed (approved):
I resolved the issues and additionally added the share button to all warnings. I also added a mechanism to adapt the warning / share behavior to be used in the hub.
